include/zephyr/drivers/stepper_motor_controller.h

```diff
@@ -0,0 +1,382 @@
+							   int32_t *position);
+
+/**
+ * @typedef stepper_motor_controller_set_position_t
+ * @brief Set the target position in stepper motor controller which should set stepper motor in
+ * motion
+ */
+typedef int32_t (*stepper_motor_controller_set_position_t)(const struct device *dev,
+							   int32_t position);
+
+/**
+ * @typedef stepper_motor_controller_enable_velocity_mode_t
+ * @brief Stepper motor controller shall activate the mode where the motor reaches a constant
+ * velocity and runs incessantly
+ */
+typedef int32_t (*stepper_motor_controller_enable_velocity_mode_t)(const struct device *dev,
+								   int32_t steps_per_second);
+
+/**
+ * @typedef stepper_motor_controller_get_stall_status_t
+ * @brief Get the status if a motor is stalled or not from stepper motor controller
+ */
+typedef int32_t (*stepper_motor_controller_get_stall_status_t)(const struct device *dev,
+							       bool *stall_status);
+
+/**
+ * @typedef stepper_motor_controller_enable_stall_detection_t
+ * @brief Enable or Disable stall detection
+ */
+typedef int32_t (*stepper_motor_controller_enable_stall_detection_t)(const struct device *dev,
+								     bool enable_stall_detection);
+
+/**
+ * @typedef stall_detection_motor_controller_set_stall_detection_threshold_t
+ * @brief Set the stall detection threshold
+ */
+typedef int32_t (*stall_detection_motor_controller_set_stall_detection_threshold_t)(
+	const struct device *dev, int32_t stall_detection_threshold);
+
+__subsystem struct stepper_motor_controller_api {
+	stepper_motor_controller_enable_t stepper_motor_controller_enable;
+	stepper_motor_controller_set_steps_to_move_t stepper_motor_controller_set_steps_to_move;
+	stepper_motor_controller_set_velocity_t stepper_motor_controller_set_velocity;
+	stepper_motor_controller_set_micro_step_res_t stepper_motor_controller_set_micro_step_res;
+#ifdef CONFIG_STEPPER_MOTOR_CONTROLLER_HAS_CONSTANT_VELOCITY_MODE
+	stepper_motor_controller_enable_velocity_mode_t
+		stepper_motor_controller_enable_velocity_mode;
+#endif /* CONFIG_STEPPER_MOTOR_CONTROLLER_HAS_CONSTANT_VELOCITY_MODE */
+#ifdef CONFIG_STEPPER_MOTOR_CONTROLLER_HAS_STALL_DETECTION
+	stepper_motor_controller_get_stall_status_t stepper_motor_controller_get_stall_status;
+
+	stepper_motor_controller_enable_stall_detection_t
+		stepper_motor_controller_enable_stall_detection;
+	stall_detection_motor_controller_set_stall_detection_threshold_t
+		stepper_motor_controller_set_stall_detection_threshold;
+#endif /* CONFIG_STEPPER_MOTOR_CONTROLLER_HAS_STALL_DETECTION */
+#ifdef CONFIG_STEPPER_MOTOR_CONTROLLER_HAS_ABSOLUTE_POSITIONING
+	stepper_motor_controller_set_reference_t stepper_motor_controller_set_reference;
+	stepper_motor_controller_get_position_t stepper_motor_controller_get_position;
+	stepper_motor_controller_set_position_t stepper_motor_controller_set_position;
+#endif /* CONFIG_STEPPER_MOTOR_CONTROLLER_HAS_ABSOLUTE_POSITIONING */
+};
```

Comment:
There's a lot of device specific stuff in here, function pointers that I suspect only applies to some specific implementation. I see you are trying to optimize this with those HAS* config option but I don't think that's a very elegant approach.

How about setting on a set of basic APIs (enable, set step, set velocty, move...) and then have a "set attribute" type of thing for everything else? See the sensors subsystem for example.

Then you could add whatever you want in the attribute list, but also have device specific ones (again, see sensor)